using EngineLayer;
using EngineLayer.GlycoSearch;
using iText.StyledXmlParser.Jsoup.Nodes;
using MassSpectrometry;
using Nett;
using NUnit.Framework;
using NUnit.Framework.Legacy;
using Omics.Modifications;
using Proteomics.ProteolyticDigestion;
using Readers;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using EngineLayer.DatabaseLoading;
using TaskLayer;
using UsefulProteomicsDatabases;

namespace Test
{
    [TestFixture]
    public class TestNOGlyco
    {
        [Test]
        public static void NandO_GlycoSearchIndividualFileFolderOutputTest()
        {
            string outputFolder = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TESTGlycoData");
            Directory.CreateDirectory(outputFolder);
            string proteinDatabase = Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\N_O_glycoWithFileSpecific\\FourMucins_NoSigPeps_FASTA.fasta");
            string spectraFileDirctory = Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\N_O_glycoWithFileSpecific");
            List<string> rawFilePaths = Directory.GetFiles(spectraFileDirctory).Where(p => p.Contains("mzML")).ToList();

            // run task
            CommonParameters commonParameters = new(dissociationType: DissociationType.HCD, ms2childScanDissociationType: DissociationType.EThcD, maxThreadsToUsePerFile: 23);

            Directory.CreateDirectory(outputFolder);
            var glycoSearchTask = new GlycoSearchTask()
            {
                CommonParameters = commonParameters,
                _glycoSearchParameters = new GlycoSearchParameters()
                {
                    OGlycanDatabasefile = "OGlycan.gdb",
                    NGlycanDatabasefile = "NGlycan_ForNoSearch.gdb",
                    GlycoSearchType = GlycoSearchType.N_O_GlycanSearch,
                    OxoniumIonFilt = true,
                    DecoyType = DecoyType.Reverse,
                    GlycoSearchTopNum = 50,
                    MaximumOGlycanAllowed = 2,
                    DoParsimony = true,
                    WriteContaminants = true,
                    WriteDecoys = true,
                    WriteIndividualFiles = true
                }
            };
            GlobalVariables.NGlycanDatabasePaths.Add(Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"NGlycan_ForNoSearch.gdb"));
            glycoSearchTask.RunTask(outputFolder, new List<DbForTask> { new DbForTask(proteinDatabase, false) }, rawFilePaths, "");

            List<string> expectedOutput = new()
            {
                "_AllProteinGroups.tsv",
                "AutoGeneratedManuscriptProse.txt",
                "results.txt",
                "no_glyco.psmtsv",
                "AllPSMs.psmtsv",
                "protein_no_glyco_localization.tsv",
                "seen_no_glyco_localization.tsv"
            };

            List<string> expectedIndividualFileOutput = new()
            {
                "2019_09_16_StcEmix_35trig_EThcD25_rep1_4999-5968_AllProteinGroups.tsv",
                "2019_09_16_StcEmix_35trig_EThcD25_rep1_4999-5968_AllPSMs.psmtsv",
                "2019_09_16_StcEmix_35trig_EThcD25_rep1_4999-5968no_glyco.psmtsv",
                "2019_09_16_StcEmix_35trig_EThcD25_rep1_4999-5968protein_no_glyco_localization.tsv",
                "2019_09_16_StcEmix_35trig_EThcD25_rep1_4999-5968seen_no_glyco_localization.tsv"
            };

            List<string> output = Directory.GetFiles(outputFolder).Select(f => Path.GetFileName(f)).ToList();
            List<string> outputFolders = Directory.GetDirectories(outputFolder).ToList();
            List<string> individualOutputFolders = Directory.GetDirectories(outputFolders.FirstOrDefault()).ToList();
            List<string> individualOutputs = Directory.GetFiles(individualOutputFolders.FirstOrDefault()).Select(f => Path.GetFileName(f)).ToList();

            CollectionAssert.IsSubsetOf(expectedOutput, output);
            CollectionAssert.IsSubsetOf(expectedIndividualFileOutput, individualOutputs);

            string[] allProteinGroups = File.ReadAllLines(Path.Combine(outputFolder, "_AllProteinGroups.tsv"));
            string[] proteinGroupFields = allProteinGroups[1].Split('\t');

            Assert.That(proteinGroupFields[0], Is.EqualTo("Q8WXI7.3"));

            Directory.Delete(outputFolder, true);
        }

        [Test]
        public static void NandO_GlycoSearch_onlyN()
        {
            string outputFolder = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TESTGlycoData");
            Directory.CreateDirectory(outputFolder);
            DbForTask db = new DbForTask(Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\Q9C0Y4.fasta"), false);
            string raw = Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\yeast_glycan_25170.mgf");
            var task = Toml.ReadFile<GlycoSearchTask>(Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\NGlycanSearchTaskconfig.toml"), MetaMorpheusTask.tomlConfig);

            task._glycoSearchParameters.GlycoSearchType = GlycoSearchType.N_O_GlycanSearch;
            task._glycoSearchParameters.NGlycanDatabasefile = "NGlycan_ForNoSearch.gdb";
            task._glycoSearchParameters.OGlycanDatabasefile = "OGlycan.gdb";
            task._glycoSearchParameters.MaximumOGlycanAllowed = 1;
            GlobalVariables.NGlycanDatabasePaths.Add(Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"NGlycan_ForNoSearch.gdb"));
            // run task
            new EverythingRunnerEngine(new List<(string, MetaMorpheusTask)> { ("Task", task) }, new List<string> { raw }, new List<DbForTask>
            {
                db
            }, Path.Combine(Environment.CurrentDirectory, @"TESTGlycoData")).Run();

            // Check the NOBoxes are built correctly
            Assert.That(GlycanBox.GlobalOGlycans.Length == 24);
            Assert.That(GlycanBox.GlobalNGlycans.Count == 10);
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p < 0) > 1)); // no NO box has more than one Nglycan (ids negative)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.NumberOfMods > task._glycoSearchParameters.MaximumOGlycanAllowed + 1 )); // no NO box has more than 2 glycans (2 O + 1N)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p >= 0) > task._glycoSearchParameters.MaximumOGlycanAllowed)); // no NO box has more than 1 Oglycans (ids non-negative)

            //For PSMs
            var allPsmPath = Path.Combine(outputFolder, "Task", "AllPSMs.psmtsv");
            List<PsmFromTsv> Psms = SpectrumMatchTsvReader.ReadPsmTsv(allPsmPath, out var errors2).ToList();
            Assert.That(errors2.Count == 0);// if we cannot find the file, we will get an error message
            Assert.That(Psms.Count == 1);
            Assert.That(Psms.First().FullSequence == "DAN[N-linked glycosylation:H5N2 on Nxt]NTQFQFTSR");

            //For GlycoPSMs
            string oGlycoPath = Path.Combine(outputFolder, "Task", "no_glyco.psmtsv");
            List<PsmFromTsv> glycoPsms = SpectrumMatchTsvReader.ReadPsmTsv(oGlycoPath, out var errors).ToList(); // the filtering (Q<0.01)
            Assert.That(errors.Count == 0);// if we cannot find the file, we will get an error message
            Assert.That(glycoPsms.Count == 1);
            Assert.That(glycoPsms.First().FullSequence == "DAN[N-linked glycosylation:H5N2 on Nxt]NTQFQFTSR");

            Directory.Delete(outputFolder, true);
        }

        [Test]
        public static void NandO_GlycoSearch_onlyO()
        {
            string outputFolder = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TESTGlycoData");
            Directory.CreateDirectory(outputFolder);
            DbForTask db = new DbForTask(Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\P16150.fasta"), false);
            string raw = Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\2019_09_16_StcEmix_35trig_EThcD25_rep1_9906.mgf");
            var task = Toml.ReadFile<GlycoSearchTask>(Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\GlycoSearchTaskconfigOGlycoTest_Run.toml"), MetaMorpheusTask.tomlConfig);

            task._glycoSearchParameters.GlycoSearchType = GlycoSearchType.N_O_GlycanSearch;
            task._glycoSearchParameters.NGlycanDatabasefile = "NGlycan_ForNoSearch.gdb";
            task._glycoSearchParameters.OGlycanDatabasefile = "OGlycan.gdb";
            task._glycoSearchParameters.MaximumOGlycanAllowed = 3;
            GlobalVariables.NGlycanDatabasePaths.Add(Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"NGlycan_ForNoSearch.gdb"));
            // run task
            new EverythingRunnerEngine(new List<(string, MetaMorpheusTask)> { ("Task", task) }, new List<string> { raw }, new List<DbForTask>
            {
                db
            }, Path.Combine(Environment.CurrentDirectory, @"TESTGlycoData")).Run();

            // Check the NOBoxes are built correctly
            Assert.That(GlycanBox.GlobalOGlycans.Length == 24);
            Assert.That(GlycanBox.GlobalNGlycans.Count == 10);
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p < 0) > 1)); // no NO box has more than one Nglycan (ids negative)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.NumberOfMods > task._glycoSearchParameters.MaximumOGlycanAllowed + 1 )); // no NO box has more than 3 glycans (2 O + 1N)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p >= 0) > task._glycoSearchParameters.MaximumOGlycanAllowed)); // no NO box has more than 3 Oglycans (ids non-negative)

            //For PSMs
            var allPsmPath = Path.Combine(outputFolder, "Task", "AllPSMs.psmtsv");
            List<PsmFromTsv> Psms = SpectrumMatchTsvReader.ReadPsmTsv(allPsmPath, out var errors2).ToList();
            Assert.That(errors2.Count == 0);// if we cannot find the file, we will get an error message
            Assert.That(Psms.Count == 1);
            Assert.That(Psms.First().FullSequence == "TTGS[O-linked glycosylation:H1N1 on S]LEPS[O-linked glycosylation:H2N2A1 on S]S[O-linked glycosylation:H2N2A1F1 on S]GASGPQVSSVK");
            Assert.That(Psms.First().Score, Is.EqualTo(13.0).Within(0.1));
            Directory.Delete(outputFolder, true);
        }

        [Test]
        public static void NandO_GlycoSearch_onlyO_Run2()
        {
            string outputFolder = Path.Combine(TestContext.CurrentContext.TestDirectory, @"TESTGlycoData");
            Directory.CreateDirectory(outputFolder);
            DbForTask db = new DbForTask(Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\P02649.fasta"), false);
            string raw = Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\181217_Fusion_(LC2)_NewObj_Serum_deSA_Jacalin_HRM_4h_ETD_HCD_DDA_mz(400_1200)_21707.mgf");
            var task = Toml.ReadFile<GlycoSearchTask>(Path.Combine(TestContext.CurrentContext.TestDirectory, @"GlycoTestData\GlycoSearchTaskconfigOGlycoTest_Run2.toml"), MetaMorpheusTask.tomlConfig);

            task._glycoSearchParameters.GlycoSearchType = GlycoSearchType.N_O_GlycanSearch;
            task._glycoSearchParameters.NGlycanDatabasefile = "NGlycan_ForNoSearch.gdb";
            task._glycoSearchParameters.OGlycanDatabasefile = "OGlycan.gdb";
            task._glycoSearchParameters.MaximumOGlycanAllowed = 1;
            GlobalVariables.NGlycanDatabasePaths.Add(Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"NGlycan_ForNoSearch.gdb"));
            // run task
            new EverythingRunnerEngine(new List<(string, MetaMorpheusTask)> { ("Task", task) }, new List<string> { raw }, new List<DbForTask>
            {
                db
            }, Path.Combine(Environment.CurrentDirectory, @"TESTGlycoData")).Run();

            // Check the NOBoxes are built correctly
            Assert.That(GlycanBox.GlobalOGlycans.Length == 24);
            Assert.That(GlycanBox.GlobalNGlycans.Count == 10);
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p < 0) > 1)); // no NO box has more than one Nglycan (ids negative)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.NumberOfMods > task._glycoSearchParameters.MaximumOGlycanAllowed + 1)); // no NO box has more than 2 glycans (1 O + 1N)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p >= 0) > task._glycoSearchParameters.MaximumOGlycanAllowed)); // no NO box has more than 1 Oglycans (ids non-negative)

            //For PSMs
            var allPsmPath = Path.Combine(outputFolder, "Task", "AllPSMs.psmtsv");
            List<PsmFromTsv> Psms = SpectrumMatchTsvReader.ReadPsmTsv(allPsmPath, out var errors2).ToList();
            Assert.That(errors2.Count == 0);// if we cannot find the file, we will get an error message
            Assert.That(Psms.Count == 1);
            Assert.That(Psms.First().FullSequence == "AAT[O-linked glycosylation:N1 on T]VGSLAGQPLQER");
            Assert.That(Psms.First().Score, Is.EqualTo(23.2).Within(0.1));
            Directory.Delete(outputFolder, true);
        }


        [Test]
        public static void TestNOGlycanBox()
        {
            // Loading the glycan databases
            string oglycanPath = Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"OGlycan.gdb");
            string nglycanPath = Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"NGlycan_NOBoxesTesting.gdb");

            // reading the O-Glycan database and storing it in GlycanBox.GlobalOGlycans
            GlycanBox.GlobalOGlycans = GlycanDatabase.LoadGlycan(oglycanPath, true, true).ToArray();
            Assert.That(GlycanBox.GlobalOGlycans.Count() == 24);
            Assert.That(GlycanBox.GlobalOGlycans.All(p => p.ModificationType == "O-linked glycosylation"));

            // reading the N-Glycan database and storing it in GlycanBox.GlobalNGlycans with negative keys
            GlycanBox.GlobalNGlycans = new Dictionary<int, Glycan>();
            var nGlycans = GlycanDatabase.LoadGlycan(nglycanPath, true, false).ToArray();
            int indexForNGlycan = -1;
            foreach (var nGlycan in nGlycans)
            {
                GlycanBox.GlobalNGlycans.Add(indexForNGlycan, nGlycan);
                indexForNGlycan--;
            }
            Assert.That(GlycanBox.GlobalNGlycans.Count == 20); // we have 20 Nglycans in the database
            Assert.That(GlycanBox.GlobalNGlycans.All(p=>p.Key < 0)); // all keys in GlobalNGlycans should be negative
            Assert.That(GlycanBox.GlobalNGlycans.All(p=>p.Value.ModificationType == "N-linked glycosylation"));

            // Building the NO_glycan boxes
            int _maxOGlycanNum = 3;
            GlycanBox.NOGlycanBoxes = GlycanBox.BuildNOGlycanBoxes(_maxOGlycanNum, false).OrderBy(p => p.Mass).ToArray();
            
            // Check the number of glycan in each box is less than or equal to max glycan number
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p=>p.NumberOfMods > _maxOGlycanNum + 1));
            // Check there is at most one N-glycan in each box (p is negative means Nglycan)
            Assert.That(!GlycanBox.NOGlycanBoxes.Any(p => p.ModIds.Count(p => p < 0) > 1));
            // Check there is no duplicated box
            var boxSet = new HashSet<string>();
            foreach (var glycanBox in GlycanBox.NOGlycanBoxes)
            {
                var key = string.Join(",", glycanBox.ModIds.OrderBy(p => p));
                Assert.That(!boxSet.Contains(key));
                boxSet.Add(key);
            }

            // Make sure some box is Nglycan only
            Assert.That(GlycanBox.NOGlycanBoxes.Any(p=>p.ModIds.Count(p=>p >= 0) == 0));
            // Check the NOBoxes with Nglycan only is equal to the number of Nglycans in the Nglycan database

            // we count by checking ModIds without any non-negative id (non-negative ids are Oglycans)
            int countForNglycanOnly_byModId = GlycanBox.NOGlycanBoxes.Count(p => p.ModIds.Count(p => p >= 0) == 0);
            Assert.That(countForNglycanOnly_byModId == GlycanBox.GlobalNGlycans.Count);
            // we count by checking OGlycanIds is null
            int countForNglycanOnly_byOGlycanIds = GlycanBox.NOGlycanBoxes.Count(p => p.OGlycanIds == null);
            Assert.That(countForNglycanOnly_byOGlycanIds == GlycanBox.GlobalNGlycans.Count);

            //Make sure some box is Oglycan only
            int countForOglycanOnly = GlycanBox.NOGlycanBoxes.Count(p => p.ModIds.Length == 1 && p.OGlycanIds != null);
            Assert.That(countForOglycanOnly == GlycanBox.GlobalOGlycans.Length);


            // Check the mass of each box is equal to the sum of the mass of glycans in the box
            foreach (var glycanBox in GlycanBox.NOGlycanBoxes)
            {
                double totalOglycanMass = glycanBox.ModIds.Where(p => p >= 0).Select(p => GlycanBox.GlobalOGlycans[p].Mass / 1E5).Sum();
                double totalNglycanMass = glycanBox.ModIds.Where(p => p < 0).Select(p => GlycanBox.GlobalNGlycans[p].Mass / 1E5).Sum();
                Assert.That(glycanBox.Mass, Is.EqualTo(totalNglycanMass + totalOglycanMass).Within(1E-4));
            }
        }

        [Test]
        public static void TestNOGlycanBox_Decoy()
        {
            // Loading the glycan databases
            string oglycanPath = Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"OGlycan.gdb");
            string nglycanPath = Path.Combine(TestContext.CurrentContext.TestDirectory, "GlycoTestData", @"NGlycan_NOBoxesTesting.gdb");

            // reading the Glycan database and storing it in GlycanBox
            GlycanBox.GlobalOGlycans = GlycanDatabase.LoadGlycan(oglycanPath, true, true).ToArray();
            GlycanBox.GlobalNGlycans = new Dictionary<int, Glycan>();
            var nGlycans = GlycanDatabase.LoadGlycan(nglycanPath, true, false).ToArray();
            int indexForNGlycan = -1;
            foreach (var nGlycan in nGlycans)
            {
                GlycanBox.GlobalNGlycans.Add(indexForNGlycan, nGlycan);
                indexForNGlycan--;
            }

            // Building the NO_glycan boxes
            int _maxOGlycanNum = 2;
            GlycanBox.NOGlycanBoxes = GlycanBox.BuildNOGlycanBoxes(_maxOGlycanNum, true).OrderBy(p => p.Mass).ToArray();
            Assert.That(GlycanBox.NOGlycanBoxes.Any(p=>p.TargetDecoy == false));
            Assert.That(GlycanBox.NOGlycanBoxes.Count(p=>!p.TargetDecoy) == GlycanBox.NOGlycanBoxes.Count(p=>p.TargetDecoy));

            foreach (var decoyBox in GlycanBox.NOGlycanBoxes.Where(p=>!p.TargetDecoy))
            {
                var theBoxWithSameComponent = GlycanBox.NOGlycanBoxes.Where(p => p.TargetDecoy).Where(p => p.ModIds.SequenceEqual(decoyBox.ModIds));
                Assert.That(theBoxWithSameComponent.Count() == 1);
                var targetBox = theBoxWithSameComponent.First();
                Assert.That(decoyBox.NGlycanId == targetBox.NGlycanId);
                Assert.That(decoyBox.OGlycanIds == targetBox.OGlycanIds);
                Assert.That(decoyBox.Mass != targetBox.Mass);
            }
        }

        [Test]
        public static void TestGetModPos()
        {
            //testing the function to get the possible modification sites on a peptide with setMod. If the N in NxS/T is modified, then this site should not be counted as a possible N-glycosylation site.

            // case 1: no modification on the peptide
            PeptideWithSetModifications pep = new PeptideWithSetModifications("ELNPTPNVEVNVSCR", null);
            string[] motifs = new string[] {"S", "T",  "Nxs", "Nxt" };
            var sites = GlycoSpectralMatch.GetPossibleModSites(pep, motifs);
            Assert.That(sites.Count() == 4 && sites.ContainsKey(4) && sites.ContainsKey(6) && sites.ContainsKey(12) && sites.ContainsKey(14));
            Assert.That(sites[4] == "Nxt" && sites[6] == "T");
            Assert.That(sites[12] == "Nxs" && sites[14] == "S");


            // case 2: with modification on NxT
            ModificationMotif.TryGetMotif("N", out ModificationMotif motif);
            Modification mod = new Modification(_originalId: "Test of N", _modificationType: "Common Fixed", _target: motif, _locationRestriction: "Anywhere.");
            var testN = new PeptideWithSetModifications("ELN[Common Fixed:Test of N]PTPNVEVNVSCR", new Dictionary<string, Modification> {  { "Test of N", mod } });
            var site_case2 = GlycoSpectralMatch.GetPossibleModSites(testN, motifs);
            Assert.That(site_case2.Count() == 3 && site_case2.ContainsKey(6) && site_case2.ContainsKey(12) && site_case2.ContainsKey(14));
            Assert.That(site_case2[6] == "T");
            Assert.That(site_case2[12] == "Nxs" && site_case2[14] == "S");

            // case 3: with modification on NxS
            var testN_2 = new PeptideWithSetModifications("ELNPTPNVEVN[Common Fixed:Test of N]VSCR", new Dictionary<string, Modification> { { "Test of N", mod } });
            var site_case3 = GlycoSpectralMatch.GetPossibleModSites(testN_2, motifs);
            Assert.That(site_case3.Count() == 3 && site_case3.ContainsKey(4) && site_case3.ContainsKey(6) && site_case3.ContainsKey(14));
            Assert.That(site_case3[4] == "Nxt" && site_case3[6] == "T");
            Assert.That(site_case3[14] == "S");
        }
    }
}
